src/ui.rs

```diff
@@ -892,4 +404,446 @@ impl<T> Snarl<T> {
+
+        r.context_menu(|ui| {
+            viewer.node_menu(node_idx, &inputs, &outputs, ui, snarl_state.scale(), self);
+        });
+
+        if !self.nodes.contains(node_idx) {
+            node_state.clear(ui.ctx());
+            // If removed
+            return response;
+        }
+
+        // Rect for node + frame margin.
+        let node_frame_rect = node_frame.total_margin().expand_rect(node_rect);
+
+        let node_ui = &mut ui.child_ui_with_id_source(
+            node_frame_rect,
+            Layout::top_down(Align::Center),
+            node_id,
+        );
+        node_ui.set_style(node_style.clone());
+
+        node_frame.show(node_ui, |ui| {
+            // Render header frame.
+            let mut header_rect = node_rect;
+
+            let mut header_frame_rect = header_frame.total_margin().expand_rect(header_rect);
+
+            // Show node's header
+            let header_ui = &mut ui.child_ui_with_id_source(
+                header_frame_rect,
+                Layout::top_down(Align::Center),
+                node_id,
+            );
+
+            header_frame.show(header_ui, |ui: &mut Ui| {
+                ui.with_layout(Layout::left_to_right(Align::Min), |ui| {
+                    if style.collapsible {
+                        let (_, r) = ui.allocate_exact_size(
+                            vec2(node_style.spacing.icon_width, node_style.spacing.icon_width),
+                            Sense::click(),
+                        );
+                        paint_default_icon(ui, openness, &r);
+
+                        if r.clicked_by(PointerButton::Primary) {
+                            // Toggle node's openness.
+                            self.open_node(node_idx, !open);
+                        }
+                    }
+
+                    ui.allocate_exact_size(header_drag_space, Sense::hover());
+
+                    viewer.show_header(node_idx, &inputs, &outputs, ui, snarl_state.scale(), self);
+
+                    header_rect.max = ui.min_rect().max;
+                });
+
+                header_frame_rect = header_frame.total_margin().expand_rect(header_rect);
+
+                ui.advance_cursor_after_rect(Rect::from_min_max(
+                    header_rect.min,
+                    pos2(
+                        f32::max(header_rect.max.x, node_rect.max.x),
+                        header_rect.min.y,
+                    ),
+                ));
+            });
+            if !self.nodes.contains(node_idx) {
+                node_state.clear(ui.ctx());
+                // If removed
+                return;
+            }
+
+            let min_pin_y = header_rect.center().y;
+
+            let input_x = header_rect.left() + header_frame.total_margin().left + pin_size * 0.5;
+
+            let output_x = f32::max(header_rect.right(), node_rect.right())
+                - header_frame.total_margin().right
+                - pin_size * 0.5;
+
+            if true {
```

Comment:
`if true` may be removed.
However I still want to skip pins UI when node is fully collapsed, so you may keep it as a reminder.